### DISCUSSION TITLE
[10/18] 이승현 / Ensemble Knowledge Guided Sub-network Search and Fine-tuning for Filter Pruning

### DISCUSSION THREAD

**Discussion by sseung0703** [sjquan/2022-Study#18] [Seminars]
[Q] 기법들 마다 없어져야하는 filter가 무엇인지를 다르게 말하고 있는 것 같은데, 혹시 전체적으로 이쪽 분야에서 모두 동의하는 바가 있거나 할까요?

**[A] Filter importance를 측정하는 방법론에서 모두가 동의하는 바는 없지만, "smaller-norm-less-importance"라는 명제가 사실이 아니라는 것에는 어느 정도 의견이 일치하고 있습니다. 당연히 가장 정확한 것은 실제 학습 및 평가하는 NAS가 될테고요.**
**그러다보니 pruning 연구자가 아닐 경우, 가장 편리한 L1-norm을 data-driven criterion을 사용하고 싶을 때는 gradients-norm 정도를 사용하는 것이 보편적입니다.**

--------

[Q] pearson correlation coefficient 관련 그래프 설명해주실 때, y축에 해당하는 loss(이름을 잊어버렸습니당)가 어떤 의미를 가지는 지 궁금합니다.

**[A] y축은 potential loss로 pruned model을 fine-tuning했을 때 loss입니다. 즉, 저희가 filter pruning을 했을 때 얻게 되는 모델의 실제 성능이 되겠지요.**
**Pruned model은 dynamic network가 아니기 때문에 제 성능을 발휘하기 위해 반드시 fine-tuning이 필요하고, 이에 따라 각 filter의 중요도도 fine-tuning되었을 때 성능, 즉, potential과 연관이 높아야 하는 것이 바람직합니다.**
**저는 이것을 표현하기 위해 random sub-network의 validation loss, potential loss, 그리고 condition number를 계산하고, pcc로 관계도를 평가한 것입니다.**

--------

[Q] 논문 Table 2.에 보면 Baseline이 프루닝 적용안한거고 앙상블 방법론이 search 랑 파인튜닝 시간다 합치면 Baseline 시간보다 GPU 시간이 큰데요. 성능 차이도 크지는 않은 거 같은데, 그럼 그냥 baseline쓰는게 나은 거 아닌가요??

**[A] 해당 실험은 Baseline의 model을 FLOPS가 50% 감량될 때까지 pruning을 수행하여 성능과 학습 시간을 확인하기 위한 것입니다.**
**물론, 실제 latency가 절반으로 줄어들지는 않겠지만, 충분히 많은 capacity를 제거한 것이며, 실제로 knowledge를 사용하지 않았을 때는 일부 configurations에서 성능 저하가 발생합니다.**
**반면, knowledge를 활용할 경우, 성능이 유지 또는 더 개선되기 때문에 충분한 의미를 가집니다.**

--------

[Q] 발표해주실 때 GBN 식의 경우에는 파라미터들에 대해서 미분하고 값을 곱해줬던 것 같은데, 논문에서는 feature map 값 (activation)에 대해서 미분값을 구하고 값을 곱해주는 scoring을 사용했던 것같았습니다. 혹시 이유가 있을까요? Channel 내에는 많은 파라미터들이 있는데, 이들에 대한 대표 score는 어떻게 계산하나요?

**[A] 이 부분을 설명드릴까 하다 너무 상세한 디테일이 될 것 같아 생략했는데 관심있게 확인해주셔서 감사합니다!**
**기존 GBN은 batch norm layer에 각 channel별로 mask의 역할을 하는 parameter를 붙입니다. 그리고 이 mask에 전달되는 gradients를 통해 loss difference와 pruning의 관계를 Taylor series로 approximation하게 됩니다.**
**이 때, 이 과정을 수식적으로 다시 정리를 해보면, 제가 논문에서 활용한 것과 같이 feature map과 그 gradients의 inner product형태로 변환할 수 있습니다. 이론적으로는 같지만 부가적인 masking을 할 필요가 없고, 모든 지점에서 scoring을 할 수 있기 때문에 다양한 활용이 가능해집니다.**
**channel내에는 많은 파라미터들을 포함하고 있지만, filter pruning은 filter 단위로 중요도를 측정하기 때문에 filter의 norm 또는 feature map 한장의 중요도 등으로 판단하게 됩니다.**

--------

[Q] GBN에서는 gradient를 필터의 중요도 스코어로 활용하게 된다고 말씀하셨는데, 투입되는 샘플의 수에 따라 각 필터들이 가지는 gradient의 값이 달라질 수 있을 것 같은데 맞나요? 맞다면, 그래디언트를 구하기위해 사용하는 샘플의 수에 대한 기준같은게 있을까요?

**[A] 투입되는 샘플의 수에 따라 gradient의 값이 달라지는 게 맞습니다. 이 때, 누적될 수록 크기가 커지는 것보다는 data에 따라 달라지게 된다는 것이 중요한 점입니다. 발표 때 말씀드렸던 것과 같이 gradients는 dataset을 통해 계산되는 것이기 때문에 그 dataset에 dependency가 생기고 더 target dataset에 맞는 pruning이 수행됩니다.**
**따라서, 데이터의 샘플 수보다는, 데이터의 분포가 target task를 얼마나 잘 표현하고 있는 가가 더 중요합니다. 가장 쉬운 방법은 모든 training dataset을 다 사용하는 것이지만, 비용 문제로 일정 부분만을 sampling하여 활용합니다. classification에서는 "각 class마다 N 장 사용한다." 는 것이 일반적인 접근법입니다.**

--------

[Q] Loss의 land scape를 구하기 위해 헤시안을 사용하셨는데, 구체적으로 어떻게 계산되는지 알 수 있는 관련 자료가 있을까요?

**[A] 발표자료의 논문을 읽어보시는 것이 가장 좋을 것 같습니다.**
```
 Li, Hao, et al. "Visualizing the loss landscape of neural nets." Advances in neural information processing systems 31 (2018).
```
**Loss landscape를 visualization하거나 smoothness를 계산하기 위해 Hessian을 계산할 때는, 계산될 방향이 중요합니다. 저는 Hessian을 계산할 때는 pre-trained network와 pruned network 사이를 계산했고, loss landscape는 그것에 orthogonal한 random direction을 추가했습니다. 무슨 말인지 제 글만 보면 이해하시기 어려울것같은데 논문을 보시면 더 이해하기 어렵지 않을거에요.**

--------

[Q] 특정 gpu에서 모델을 사용하고자 할때 pruning 시에 channel 수를 몇 배수로 자르는 것이 좋은지 판단하는 기준이 있을지 궁금합니다. (gpu 제품마다 기준이 달라질 수 있는지 궁금합니다!)

**[A] 특별한 기준은 없습니다만, 당연하게도 낮을수록 더 상세한 search가 가능하기 때문에 성능이 개선될 가능성이 있습니다. GPU마다 CUDA core 수도 다르고, 기반 아키텍처가 다르기 때문에, 최적의 단위가 있을 것으로 생각되지만 저도 정확하게는 알기 어렵습니다.**
**다만 발표 때도 질문이 나왔지만, NPU를 설계하시는 분들은 어느 정도 아실것이라 문의하시면 좋을 것 같습니다.**

--------

[Q] loss landscape을 visualize 해서 보여주셨는데 해당 visualization을 어떻게 하신건지 궁금합니다.

**[A] 위의 답변 확인부탁드립니다! 만약 visualization tool이 궁금하신 것이라면 "Paraview"라는 tool을 사용했습니다.**

--------

[Q] Pruning 관련 연구하시면서 inference time을 측정해보신 경험이 있으실까요? 있다면 어떤 framework/library/device를 사용하셨을까요?

**[A] 거의 Tensorflow를 사용하고 있고, JAX나 Pytorch도 간혹 사용하고 있습니다. 대부분의 경우에서 Python에서 GPU에 올려서 측정하고 있고, NPU도 가끔 사용해보았지만 주로 활용하지는 않아서 정확한 답변을 드리기는 어려울 것 같네요.**
**대부분의 경우에서 FLOPS 감량보다 낮은 inference time 개선이 이루어집니다. 저는 러프하게 그 절반 정도만 감량된다고 기대하고 있습니다.**

--------

[Q] 많은 pruning work들이 FLOPs / 파라미터 수로 성능을 compare하는 듯한데 structured pruning이라 실제로도 FLOPs 계산이 inference때 throughput 혹은 latency와 경향이 일치하는지 궁금합니다. (e.g. GPU에서..?) Quantization 쪽은 실제 latency도 점점 중요하게 생각하는 것 같아서, pruning 쪽도 그런 움직임이 있는지 궁금합니다 :)

**[A] 위에 답변 먼저 확인 부탁드립니다!**
**어느 정도 batch 크기가 크다면 FLOPS 감량과 throughtput의 감량은 거의 비례합니다. 하지만 latency의 경우 잘 따라가지 못하는 모습을 보이고, 이에 따라 latency를 고려하려는 움직임도 있는 것으로 알고 있습니다. 예를 들면 NAS의 reward에 latency를 넣는 등의 방식이 있습니다.**
**하지만, 측정한다 하더라도 benchmarking을 하기가 쉽지 않기 때문에 비교하시기는 어려울 것 같습니다.**

--------

[Q] Image-classification 말고 다른 task (object detection이나 segmentation)쪽에도 적용했을 경우 결과가 어느정도 보장이 될지 궁금합니다. 생각해보신적이 있으실까요?

**[A] 최근 대부분의 기법들은 detection까지는 실험에 넣고 있어서 확인해보실 수 있습니다.**
**저는 아무래도 필수 실험인 ImageNet을 하느라 바빠서 제대로 하지 못했는데 이번 논문에는 싣기 위해 노력 중입니다...**
**경향성은 비슷하지만 성능이 유의미하게 떨어지는 지점이 더 높습니다.**
**그래서 classification과 달리 50% 정도 pruning하는 경우는 드물고, 하더라도 어느 정도 성능 손실이 발생합니다.**

--------

[Q] Hessian이 무거운 연산임에는 분명하지만 꽤나 효과를 볼 수 있는 방법으로 쓰일 수 있지 않을까 싶습니다. 효과를 볼 수 있다라고 생각하는 저의 의견에 대해서 발표자분의 생각을 꼭 한 번 들어보고 싶습니다! (저는 아무래도 Hessian에 대해서는 몇 편의 논문을 읽기만 해서 경험적인 이해도가 낮은 편이라서요.)

**[A] 제가 Hessian이 쓰기 어렵다고 한 것은 그 비용 때문입니다. 그 정도의 비용을 지불해야할 정도라면 그냥 full-dataset을 밀어넣어서 generality를 확보다는 게 낫지 않을까 생각이 들게 되거든요.** 
**물론 제가 추구하는 방향과 달라서 제대로 실험하지 않았기 때문에 bias가 있다고 생각은 들고, 실험을 통해 비교하는 것도 좋은 접근법이 될 것 같습니다. 만약 단순히 dataset의 크기를 늘리는 것으로 도달하지 못하는 성능 개선이 있을 수 있다면 충분히 의미있는 결과라고 생각합니다.**
**다만, 그 결과가 모두를 위한 방법론이 되지는 못하겠지만요...**

--------

[Q] (12)번 fine tuning loss에서 contrastive learning 기법으로 오버피팅을 방지하셨다고 했는데 그 부분이 잘 이해가 안되서 질문드리고 싶습니다. contrastive learning에서 negative sample은 꼭 필요가 없는 부분인가요? 아니면 수식에서 그냥 생략된 부분인가요? 

**몇 년전 부터 negative sample없이도 학습이 가능한 contrastive learning 방법들이 많이 제시가 되고 있습니다. 아래 대표적인 논문 공유드립니다.**
```
    Chen, Ting, et al. "A simple framework for contrastive learning of visual representations." International conference on machine learning. PMLR, 2020.
    Grill, Jean-Bastien, et al. "Bootstrap your own latent-a new approach to self-supervised learning." Advances in neural information processing systems 33 (2020): 21271-21284.

```

--------

[Q] supernet의 dynamic inference는 무엇을 의미하나요

**[A] Dynammic network는 그 목적에 따라 network의 구조를 dynammic하게 변경하면서 inference가 가능한 모델을 의미합니다. 가장 대표적인 방법으로 어려운 data가 들어올 경우 무거운 network, 쉬운 data가 들어올 경우 가벼운 network 방향으로 forwarding하는 방법이 있습니다.** 
**Dynammic network는 일반적으로 특별한 학습 방식을 요구하고 있고, 이를 도입하여 filter pruning에서 활용하는 것이 slimmable network에서 제안한 supernet입니다. 이 network는 문제의 난이도 책정은 하지 못하지만, filter수를 dynammic하게 slicing하면서 forwarding을 할 수 있고, 그 결과 추가적인 학습없이 sub-network evaluation이 가능해집니다. 이로 인해 search cost가 감소하는 것이고요**

--------

[Q] 다른분이 질문 주신건에 대한 추가 문의 인데요.  원래 질문이, 결국 프루닝한 모델 vs 프루닝한 모델을 따라서 새로 학습한 모델 이런 질문을 주셨던걸로 기억합니다.  후자의 경우, 이것을 경량화한 모델이라고 인정해 줄까요..? 답이 없는 문제일수도 있겠습니다만 ㅎㅎ;

**[A] 세중님께서 첫 intro때 간단하게 언급을 하신 논문인 lottery ticket theory에서 후자의 방법론을 사용합니다. 다양한 관점을 제시한 것도 사실이지만 요즘 pruning 기법들은 그 중 기법적으로 활용될 수 있는 concept만 채용해서 활용하고 있습니다.**
**애매하게 학습되어 잘못된 local minima에 빠진 network를 다시 fine-tuning하는 것보다 차라리 pruned network의 filter들을 initialization되었을 시점으로 다시 roll-back한 후 처음부터 학습하는 것이 좋다! 라는 concept인데 실제로 생각보다 높은 성능이 나옵니다.**
**하지만 제 생각이 이 가정이 맞으려면 pre-trained network가 애매하게 학습되어 있다는 전제가 있어야하고, 직접 시도하기 어려울 정도의 비용을 들여 학습된 모델을 받아 활용할 경우 그다지 좋은 선택지는 아닌 것 같습니다.**

--------

[Q] 최종 테이블에 표시된 search 시간 외에, 실제 NAS를 통해 찾은 pruning 된 모델이 inference time 등에서 얼마나 이득이 있는지 궁금합니다.

**[A] 위의 inference time관련 답변들을 확인해주시면 좋을 것 같습니다.**
**요약해드리면, inference time의 이점은 분명히 존재하지만 그 정도는 기대하는 것보다는 낮을 수 있으며 device에 dependency가 있습니다.**

--------

[Q] Filter Pruning을 진행한 후, 실제 코드레벨에서 연산이 어떻게 진행되나요? Weight Mask를 주는 방식인것인지, 아니면 Weigth를 Hooking하고 핸들링을 진행하는건지...다른 별도의 방법이 있는것인지 궁금합니다! 

**[A] 이 부분은 참 중요하지만 간단하게 답변하기 어려운 내용이네요...**
**Filter pruning을 마친 후에는 pruned 되지않은 filter만 indexing하여 새로운 tensor를 만들게 됩니다. 다양한 구현방법이 있겠지만 제가 구현한 방식은 아래와 같습니다.**
```
    if hasattr(layer, 'in_mask'):
        tensor = tensor[..., layer.in_mask.get_mask().numpy().reshape(-1).astype(bool),:]
    if hasattr(layer, 'out_mask'):
        tensor = tensor[..., layer.out_mask.get_mask().numpy().reshape(-1).astype(bool)]
```
**Tensorflow의 parameter 형식은 [Fh,Fw,Ci,Co]의 형식을 가집니다. 따라서 input feature map을 생성한 filter의 mask를 통해 'Ci' 를 index해주고, 같은 방식으로 'Co'도 masking됩니다.**
**결국 저희가 다루는 모델은 발표 때도 설명드린 것과 같이 실제적으로 작아진 모델이 됩니다.**

--------

[Q] "teacher 결과값들을 memory bank로 저장한다는 개념이 헷갈리네요. 이걸 pth 형태로 다 저장해뒀다가 앙상블 때 사용한다고 이해했는데 맞는건가요? 그럼 이 저장된 모델들을 teacher assistant로 보면 되는게 맞는지 궁금합니다."

**[A] 네 맞습니다. 저는 메모리 효율정을 위해 numpy memmap으로 저장하여 활용했습니다. 아무래도 ImageNet의 memory bank의 크기는 수십기가 정도가 되어서 RAM에 전부 올리기는 어렵거든요.**
**이와 같은 접근법은, computational cost를 memory cost로 전환한 것이라고 생각될수도 있고, 실제로 한 리뷰어가 이를 지적하기도 했습니다. 하지만 둘 중 어떤 것이 더 경제적인 관점에서 비용이 큰가를 고려했을 때 memory bank를 구성하는 것이 훨신 효율적이라고 생각합니다.**
